### PR TITLE
feat: ensure user has a single ledger device connected when pairing

### LINKED ISSUE #6970: Multiple different ledgers can connect and sign against same `deviceId`
### Overview

It's been raised that in the WIP account management feature, it's currently possible to connect multiple different ledger devices to the app for different chains, and then attempt to sign transactions with the wrong device.

We need to disambiguate this either by blocking the ability to connect multiple devices, or by segregating them based on the bus ID.

### References and additional details

Original identification of issue:
https://discord.com/channels/554694662431178782/1242516828136407110/1242934957001670846

### Acceptance Criteria

Users are unable to sign transactions for the wrong ledger device when multiple are connected at the same time
If we do support mutiple ledger devices connected concurrently, we disambiguate which device and account is connected from.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
# Description

Adds check when pairing ledger to ensure the user only has a single device connected.
- If the user has 0 devices connected, the pairing modal waits for user to connect aa device
- If the user has more than 1 device connected, the pairing modal waits for the user to disconnect all but 1 device

# Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

# Issue (if applicable)

closes #6970

# Risk
> High Risk PRs Require 2 approvals

Low risk. This resolves an issue described in #6970 but does not add risk to the existing implementation.

> What protocols, transaction types or contract interactions might be affected by this PR?

# Testing

### Definitions:
* Paired - the device is authorized in chrome to interact with the app (see screenshot below)
* Connected - the device is plugged into your computer
* Unlocked - the device passkey is entered and is ready to trade

### User connected (and unlocked) multiple of the exact same device
1. Connect and unlock 2 or more identical ledgers (e.g 2 model Ss)
2. Open the app and connect ledger wallet
3. Click "Pair new device"
4. Chrome will open a dialog box and ask to "Pair" a device - pair one of them
5. Since both devices are identical and paired, the app doesnt know which one to use, so an error message will appear telling you to disconnect all but one device

### User paired different ledger devices previously, and now has multiple connected (and unlocked) now
1. Go through the user flow of connecting a ledger device fully
2. Disconnect wallet from app, and unplug the ledger
3. Do it again on a different ledger device of a different model
4. Disconnect wallet from app, but keep it plugged in
5. Plug in and unlock the other ledger 
    - there should now be 2 different ledger models connected and unlocked
    - go to the app tab, in the top-right of the screen to the left of the chrome web page url, click the info 'i' icon -  you should see 2 ledgers paired with the page
7. Now try to connect one of the ledgers. Since both devices are paired, the app doesnt know which one to use, so an error message will appear telling you to disconnect all but one device

Paired devices:
image

# Screenshots (if applicable)

Connect ledger with fresh cache, no ledger device currently connected, then connect a ledger device to proceed:

https://github.com/shapeshift/web/assets/125113430/dba96118-fd13-4a64-abc9-2f3b015fd2ed

Existing device previously connected, no ledger device currently connected, then connect a ledger device to proceed:

https://github.com/shapeshift/web/assets/125113430/1c4abe32-df9c-483d-8d37-7c26a51abd7a

Existing device previously connect, 2 ledgers currently connected, unplug the new one and proceed:

https://github.com/shapeshift/web/assets/125113430/7a288875-9fe8-4b20-9287-6dfff5cc47f4

Existing device previously connected, 2 ledgers currently connected, click "pair new device", clears cache, notifies user that only 1 device can be connected, disconnect old device and proceed:

https://github.com/shapeshift/web/assets/125113430/9c752b0f-cc4d-46fb-b845-024b8c21d5ad

Connect ledger with fresh cache, 2 ledgers currently connected, notifies user that only 1 device can be connected, then disconnect 1 device and proceed:

https://github.com/shapeshift/web/assets/125113430/7114120f-2024-4623-b19c-d6b81b53c51e

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Waiting for testing with multiple ledger devices

File: src/context/WalletProvider/Ledger/components/Connect.tsx [hunk 1]
```diff
@@ -57,8 +65,26 @@ export const LedgerConnect = ({ history }: LedgerSetupProps) => {
         // Remove all provider event listeners from previously connected wallets
         await removeAccountsAndChainListeners()
 
+        // Pair the device, which gets approval from the browser to communicate with the Ledger USB device
         const wallet = await adapter.pairDevice()
 
+        // Check the number of connected devices. Needs a device to be paired for this to work (see above)
+        const numDevices = await handleCheckNumDevices()
```
Comment (gomesalexandre) [inline src/context/WalletProvider/Ledger/components/Connect.tsx:72]: q: Is the intent to have this displayed *before* "Pair new/existing device" click, like in these two recordings?

> At least 1 device check:

https://github.com/shapeshift/web/assets/125113430/8b51ac2c-d5bf-4bb6-a9da-beef638f9472

> At most 1 device check:

https://github.com/shapeshift/web/assets/125113430/915c4b05-7e5b-4ba6-ac38-c21675ceb829

**Review by gomesalexandre — COMMENTED** [review]
Tested locally, core logic looks sane, though I wasn't able to get this to work similar to the recordings.

- No device *ever* connected, including now:

https://jam.dev/c/be3b85e9-a5fd-4095-9bbb-0fd57ca2e91c

- No device *ever* connected but Ledger now connected:

https://jam.dev/c/ddfe5fa8-041f-4cb3-b3da-fa446edb2625

- Reconnect and "Pair existing device" flow - Ledger unpaired physically *and* using Chrome settings

I wasn't able to get to the "No Ledger connected" flow here

- Reconnect and "Pair existing device" flow - Ledger unpaired physically *and* using Chrome settings

https://jam.dev/c/8517d95d-0ecd-4cdf-aece-3c651dfbb555

- Two Ledgers - unable to test

File: src/context/WalletProvider/Ledger/components/Connect.tsx (hunk 1 above)
Comment (woodenfurniture) [reply src/context/WalletProvider/Ledger/components/Connect.tsx:72]: Correct yeah, the idea is that you cannot pair any ledger device until exactly one is connected

Comment (woodenfurniture) [reply src/context/WalletProvider/Ledger/components/Connect.tsx:72]: Correction to my above reply (this PR was authored a long time ago), the flow has changed as the browser needs to pair webusb before requesting devices. This was picked up during ops testing where the browser needs an existing pairing with a device before we can make requests to determine the number of connected devices.

I'll upload updated screen recordings.

**Comment by woodenfurniture** [comment]
@gomesalexandre updated user flows reflecting latest changes:

Connect ledger with fresh cache, no ledger device currently connected, then connect a ledger device to proceed:

https://github.com/shapeshift/web/assets/125113430/dba96118-fd13-4a64-abc9-2f3b015fd2ed

Existing device previously connected, no ledger device currently connected, then connect a ledger device to proceed:

https://github.com/shapeshift/web/assets/125113430/1c4abe32-df9c-483d-8d37-7c26a51abd7a

Existing device previously connect, 2 ledgers currently connected, unplug the new one and proceed:

https://github.com/shapeshift/web/assets/125113430/7a288875-9fe8-4b20-9287-6dfff5cc47f4

Existing device previously connected, 2 ledgers currently connected, click "pair new device", clears cache, notifies user that only 1 device can be connected, disconnect old device and proceed:

https://github.com/shapeshift/web/assets/125113430/9c752b0f-cc4d-46fb-b845-024b8c21d5ad

Connect ledger with fresh cache, 2 ledgers currently connected, notifies user that only 1 device can be connected, then disconnect 1 device and proceed:

https://github.com/shapeshift/web/assets/125113430/7114120f-2024-4623-b19c-d6b81b53c51e

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Retested locally:

- Connect ledger with fresh cache, no ledger device currently connected, then connect a ledger device to proceed:

https://jam.dev/c/e4583eb1-dcbc-44f7-8b88-c97685eb3b91

@woodenfurniture similar to the previous testing, I'm unable to get to the same state as your latest recordings - the error is showing, although only as a flash, and pretty much instantly switches to the generic "Error" screen

- Existing device previously connected, no ledger device currently connected, then connect a ledger device to proceed:

https://jam.dev/c/d3669b3b-c1cc-411d-a570-1a021d98d4a7

Same issue here re: flash

- Existing device previously connect, 2 ledgers currently connected, unplug the new one and proceed:

https://jam.dev/c/7b1e4319-7d32-48f2-a325-827a9d740804
https://jam.dev/c/2742cc92-22ee-43fb-83a5-8b4d66c9d70b

First Jam is same tab, second is after closing and reopening the tab - I was never able to get to the second Ledger pair prompt in both cases

- Connect ledger with fresh cache, 2 ledgers currently connected, notifies user that only 1 device can be connected, then disconnect 1 device and proceed:

https://jam.dev/c/386aa087-9b87-4691-9a06-9ea8a068d776

Ditto, two Ledgers connected weren't detected

**Comment by woodenfurniture** [comment]
> Retested locally, I wasn't able to get to the same states as latest recordings 🤔
> 
> * Connect ledger with fresh cache, no ledger device currently connected, then connect a ledger device to proceed:
> 
> https://jam.dev/c/e4583eb1-dcbc-44f7-8b88-c97685eb3b91
> 
> @woodenfurniture similar to the previous testing, I'm unable to get to the same state as your latest recordings - the error is showing, although only as a flash, and pretty much instantly switches to the generic "Error" screen
> 
> * Existing device previously connected, no ledger device currently connected, then connect a ledger device to proceed:
> 
> https://jam.dev/c/d3669b3b-c1cc-411d-a570-1a021d98d4a7
> 
> Same issue here re: flash
> 
> * Existing device previously connect, 2 ledgers currently connected, unplug the new one and proceed:
> 
> https://jam.dev/c/7b1e4319-7d32-48f2-a325-827a9d740804 https://jam.dev/c/2742cc92-22ee-43fb-83a5-8b4d66c9d70b
> 
> First Jam is same tab, second is after closing and reopening the tab - I was never able to get to the second Ledger pair prompt in both cases
> 
> * Connect ledger with fresh cache, 2 ledgers currently connected, notifies user that only 1 device can be connected, then disconnect 1 device and proceed:
> 
> https://jam.dev/c/386aa087-9b87-4691-9a06-9ea8a068d776
> 
> Ditto, two Ledgers connected weren't detected

@gomesalexandre my push yesterday failed it seems, please try again? 🙏
image

File: src/assets/translations/en/main.json [hunk 2]
```diff
@@ -1443,7 +1443,9 @@
         "connectFailure": "Unable to connect Ledger wallet",
         "unknown": "An unexpected error occurred communicating with Ledger wallet",
         "appNotOpen": "%{app} not open on Ledger",
-        "appNotOpenDescription": "Please open the %{app} app on your device and try again."
+        "appNotOpenDescription": "Please open the %{app} app on your device and try again.",
+        "noDeviceConnected": "No Ledger device connected",
```
Comment (gomesalexandre) [inline src/assets/translations/en/main.json:1447]: Note @woodenfurniture - I'm unable to get to this one but only "Could not initialize device" error state, see Jams

**Review by gomesalexandre — COMMENTED** [review]
Retested locally, looking much better now for most of cases!
Although still unable to repro the previously recorded 2 Ledger case @woodenfurniture, see last Jam:

- No device ever connected, including now:

https://jam.dev/c/dba9d83a-eb84-4b17-a895-9824a5f250c2

- No device ever connected but Ledger now connected:

https://jam.dev/c/6b3dc46b-fa7e-4c76-bbc0-4374c38a85d8

- Reconnect and "Pair existing device" / "Pair New Device" flow - Ledger unpaired physically and using Chrome settings

https://jam.dev/c/85de1bc1-d24b-45b7-961d-e1709baf760d

- 2 Ledgers connected

https://jam.dev/c/5da680b1-5dd6-44f0-be7d-a11f61ae855e

File: src/assets/translations/en/main.json (hunk 2 above)
Comment (woodenfurniture) [reply src/assets/translations/en/main.json:1447]: Correct yeah , the new flow means the lack of a device is caught by webusb as an initialization error like in the first vid below, so this case is more of a theoretical one than anything. Make more sense to have this case covered than implicitly return assuming upstream has it handled
https://github.com/shapeshift/web/pull/6986#issuecomment-2187640161

**Comment by woodenfurniture** [comment]
> Retested locally, looking much better now for most of cases! Although still unable to repro the previously recorded 2 Ledger case @woodenfurniture, see last Jam:
> 
> * No device ever connected, including now:
> 
> https://jam.dev/c/dba9d83a-eb84-4b17-a895-9824a5f250c2
> 
> * No device ever connected but Ledger now connected:
> 
> https://jam.dev/c/6b3dc46b-fa7e-4c76-bbc0-4374c38a85d8
> 
> * Reconnect and "Pair existing device" / "Pair New Device" flow - Ledger unpaired physically and using Chrome settings
> 
> https://jam.dev/c/85de1bc1-d24b-45b7-961d-e1709baf760d
> 
> * 2 Ledgers connected
> 
> https://jam.dev/c/5da680b1-5dd6-44f0-be7d-a11f61ae855e

@gomesalexandre i cannot reproduce the issue and I'm getting the correct error message in th UI, can we pile and debug your local when you're online? I'm wondering if its the ledger device types you have connected, as both of mine are nano S rather than 2 different models

image

**Review by gomesalexandre — APPROVED** [review]
Retested again, actually confirmed happy this time 🎉 

- Two *same model* (e.g Nano S) devices connected

Haven't been able to test it, but from the debugging we've done with @woodenfurniture, alongside with the screenshots, can attest this would be caught, as one device "paired" but two connected really means two paired (all same-model Ledgers share the same device ID)

- Two *diff model* (e.g Nano S and Nano X) devices connected, but only one paired

This is where I thought I was unable to test it, but it's in fact happy - the app will automagically use the previously paired Ledger model. If the user wishes to change their paired Ledger, they would have to do so in Chrome settings

- Two *diff model* (e.g Nano S and Nano X) devices connected, two paired

It's a bit weird to get in this flow when testing, but absolutely valid as a user:

1. Pair one device model in the app
2. Disconnect it physically and connect your other device model physically, then connect it in the app
3. In your Chrome settings, you'll now have two devices paired
4. Refresh your app again, this time with *both* models physically connected
5. Notice you see the error message

**Pull request merged**